az login: ValueError: allow_broker=True is only supported in PublicClientApplication

<!--- 🛑 Please check existing issues first before continuing: https://github.com/Azure/azure-cli/issues --->

### **This is autogenerated. Please review and update as needed.**

## Describe the bug

**Command Name**
`az login`

**Errors:**
```
% az login --service-principal -u $spnid -p $spnsecret --tenant $tenantid
The command failed with an unexpected error. Here is the traceback:
allow_broker=True is only supported in PublicClientApplication
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/azure-cli/2.47.0/libexec/lib/python3.10/site-packages/knack/cli.py", line 233, in invoke
    cmd_result = self.invocation.execute(args)
  File "/opt/homebrew/Cellar/azure-cli/2.47.0/libexec/lib/python3.10/site-packages/azure/cli/core/commands/__init__.py", line 663, in execute
    raise ex
  File "/opt/homebrew/Cellar/azure-cli/2.47.0/libexec/lib/python3.10/site-packages/azure/cli/core/commands/__init__.py", line 726, in _run_jobs_serially
    results.append(self._run_job(expanded_arg, cmd_copy))
  File "/opt/homebrew/Cellar/azure-cli/2.47.0/libexec/lib/python3.10/site-packages/azure/cli/core/commands/__init__.py", line 697, in _run_job
    result = cmd_copy(params)
  File "/opt/homebrew/Cellar/azure-cli/2.47.0/libexec/lib/python3.10/site-packages/azure/cli/core/commands/__init__.py", line 333, in __call__
    return self.handler(*args, **kwargs)
  File "/opt/homebrew/Cellar/azure-cli/2.47.0/libexec/lib/python3.10/site-packages/azure/cli/core/commands/command_operation.py", line 121, in handler
    return op(**command_args)
  File "/opt/homebrew/Cellar/azure-cli/2.47.0/libexec/lib/python3.10/site-packages/azure/cli/command_modules/profile/custom.py", line 139, in login
    subscriptions = profile.login(
  File "/opt/homebrew/Cellar/azure-cli/2.47.0/libexec/lib/python3.10/site-packages/azure/cli/core/_profile.py", line 159, in login
    identity.login_with_service_principal(username, password, scopes=scopes)
  File "/opt/homebrew/Cellar/azure-cli/2.47.0/libexec/lib/python3.10/site-packages/azure/cli/core/auth/identity.py", line 182, in login_with_service_principal
    cred = ServicePrincipalCredential(sp_auth, **self._msal_app_kwargs)
  File "/opt/homebrew/Cellar/azure-cli/2.47.0/libexec/lib/python3.10/site-packages/azure/cli/core/auth/msal_authentication.py", line 132, in __init__
    super().__init__(service_principal_auth.client_id, client_credential=client_credential, **kwargs)
  File "/opt/homebrew/Cellar/azure-cli/2.47.0/libexec/lib/python3.10/site-packages/msal/application.py", line 555, in __init__
    raise ValueError("allow_broker=True is only supported in PublicClientApplication")
ValueError: allow_broker=True is only supported in PublicClientApplication
To check existing issues, please visit: https://github.com/Azure/azure-cli/issues
To open a new issue, please run `az feedback````

## To Reproduce:
Steps to reproduce the behavior. Note that argument values have been redacted, as they may contain sensitive information.

az login --service-principal -u $spnid -p $spnsecret --tenant $tenantid

## Expected Behavior

## Environment Summary
```
macOS-13.2.1-arm64-arm-64bit, Darwin 22.3.0
Python 3.10.10
Installer: HOMEBREW

azure-cli 2.47.0

Extensions:
azure-devops 0.26.0
log-analytics-solution 0.1.1
ssh 1.1.5

Dependencies:
msal 1.20.0
azure-mgmt-resource 22.0.0
```
## Additional Context

<!--Please don't remove this:-->
<!--auto-generated-->

Having the same issue:

(.venv) PS C:\python> az feedback
The command failed with an unexpected error. Here is the traceback:
Expected 1 module to load starting with 'azext_': got []
Traceback (most recent call last):
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\knack/cli.py", line 233, in invoke
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/__init__.py", line 517, in execute
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/__init__.py", line 426, in load_command_table
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/__init__.py", line 320, in _update_command_table_from_extensions
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/__init__.py", line 307, in _filter_modname
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/extension/__init__.py", line 308, in get_extension_modname
AssertionError: Expected 1 module to load starting with 'azext_': got []
To check existing issues, please visit: https://github.com/Azure/azure-cli/issues
To open a new issue, please run `az feedback`

@anelliaf, your error `AssertionError: Expected 1 module to load starting with 'azext': got []` is different from the `ValueError: allow_broker=True is only supported in PublicClientApplication` error. Your error is tracked by https://github.com/Azure/azure-cli/issues/25473.

--- a/src/azure-cli-core/azure/cli/core/auth/identity.py
+++ b/src/azure-cli-core/azure/cli/core/auth/identity.py
@@ -85,7 +85,7 @@ class Identity:  # pylint: disable=too-many-instance-attributes
 
     @property
     def _msal_app_kwargs(self):
-        """kwargs for creating UserCredential or ServicePrincipalCredential.
+        """kwargs for creating ClientApplication (including its subclass ConfidentialClientApplication).
         MSAL token cache and HTTP cache are lazily created.
         """
         if not Identity._msal_token_cache:
@@ -98,10 +98,15 @@ class Identity:  # pylint: disable=too-many-instance-attributes
             "authority": self._msal_authority,
             "token_cache": Identity._msal_token_cache,
             "http_cache": Identity._msal_http_cache,
-            "allow_broker": self._allow_broker,
             # CP1 means we can handle claims challenges (CAE)
             "client_capabilities": None if "AZURE_IDENTITY_DISABLE_CP1" in os.environ else ["CP1"]
         }
+
+    @property
+    def _msal_public_app_kwargs(self):
+        """kwargs for creating PublicClientApplication."""
+        # allow_broker can only be used on PublicClientApplication.
+        return {**self._msal_app_kwargs, "allow_broker": self._allow_broker}
 
     @property
     def _msal_app(self):
@@ -109,7 +114,7 @@ class Identity:  # pylint: disable=too-many-instance-attributes
         The instance is lazily created.
         """
         if not self._msal_app_instance:
-            self._msal_app_instance = PublicClientApplication(self.client_id, **self._msal_app_kwargs)
+            self._msal_app_instance = PublicClientApplication(self.client_id, **self._msal_public_app_kwargs)
         return self._msal_app_instance
 
     def _load_msal_token_cache(self):
@@ -222,7 +227,7 @@ class Identity:  # pylint: disable=too-many-instance-attributes
         return accounts
 
     def get_user_credential(self, username):
-        return UserCredential(self.client_id, username, **self._msal_app_kwargs)
+        return UserCredential(self.client_id, username, **self._msal_public_app_kwargs)
 
     def get_service_principal_credential(self, client_id):
         entry = self._service_principal_store.load_entry(client_id, self.tenant_id)

